src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Runtime.ConstrainedExecution;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO.Enumeration
+{
+    public unsafe abstract partial class FileSystemEnumerator<TResult> : CriticalFinalizerObject, IEnumerator<TResult>
+    {
+        private readonly string _originalRootDirectory;
+        private readonly string _rootDirectory;
+        private readonly EnumerationOptions _options;
+
+        private object _lock = new object();
+
+        private string _currentPath;
+        private SafeDirectoryHandle _directoryHandle;
+        private bool _lastEntryFound;
+        private Queue<(SafeDirectoryHandle Handle, string Path)> _pending;
+
+        private Interop.Sys.DirectoryEntry _entry;
+        private TResult _current;
+
+        /// <summary>
+        /// Encapsulates a find operation.
+        /// </summary>
+        /// <param name="directory">The directory to search in.</param>
+        /// <param name="options">Enumeration options to use.</param>
+        public FileSystemEnumerator(string directory, EnumerationOptions options = null)
+        {
+            _originalRootDirectory = directory ?? throw new ArgumentNullException(nameof(directory));
+            _rootDirectory = Path.GetFullPath(directory);
+            _options = options ?? EnumerationOptions.Default;
+
+            // We need to initialize the directory handle up front to ensure
+            // we immediately throw IO exceptions for missing directory/etc.
+            _directoryHandle = CreateDirectoryHandle(_rootDirectory);
+            if (_directoryHandle == null)
+                _lastEntryFound = true;
+
+            _currentPath = _rootDirectory;
+        }
+
+        private static SafeDirectoryHandle CreateDirectoryHandle(string path)
```

Comment:
You mentioned that SafeHandle is costly. Is that something you have a plan to avoid?

Reply:
Yes. That is part of the pending work.